File: binary_search/boj/p1654.py
from sys import stdin

k, n = map(int, stdin.readline().split())
lan = [int(stdin.readline()) for _ in range(k)]
start, end = 1, max(lan)
result = 0

while start <= end:
    sum = 0
    mid = (start + end) // 2

    for i in lan:
        if i >= mid:
            sum += (i // mid)
    
    if sum < n:
        end = mid - 1
    else:
        result = mid
        start = mid + 1

print(result)

File: binary_search/boj/P1654.java
package binary_search.boj;

import java.util.Arrays;
import java.util.Scanner;

public class P1654 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int k = sc.nextInt();
        int n = sc.nextInt();

        int[] len = new int[k];

        for (int i = 0; i < k; i++) {
            len[i] = sc.nextInt();
        }

        sc.close();

        Arrays.sort(len);

        long start = 1;
        long end = len[k-1]; 
        long sum = 0;
        long mid = 0;
        long result = 0;

        while (start <= end) {
            sum = 0;
            mid = (start + end) / 2;

            for (long i : len) {
                if (i >= mid) {
                    sum += i / mid;
                }
            }

            if (sum >= n) {
                result = mid;
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        System.out.println(result);
    }
}
